### PR TITLE
Add CORS error detection and convert the TypeError to AJAXError

### PR DESCRIPTION
**Pull request by jonathanlurie**
## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with a compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [ ] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. **Read bellow**
 - [x] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

## Description
This PR addresses how failed HTTP requests due to CORS policy are handled. So far, if a request fails because of CORS, it does not return a response, instead, it makes the browser throw a generic TypeError ( [see here](https://github.com/maplibre/maplibre-gl-js/blob/v5.0.0-pre.1/src/util/ajax.ts#L158) ). As a result, a map instance subscribing to the `"error"` event would catch a generic error that does not include the faulty URL. It is better to handle a CORS error with an AJAXError, that contains information about the URL and the cause of the actual problem.  

This is especially useful in the context of loading a style.json so that if a given style is not reachable because of CORS, it is possible to catch this error with the same efficiency as if fetching a style.json would yield an HTTP 5xx status code, since in both cases the request would fails due to server configuration.

## Before / After
With this minimalist snippet:
```js
const map = new maplibregl.Map({
  container: 'map',
  style: "https://example.com/style.json",
});

map.on("error", (err) => {
  console.log(">>> ", err);
}) 
```
Here is the console output,
- **BEFORE:**
Screenshot 2024-10-11 at 09 59 01

- **AFTER:**
Screenshot 2024-10-11 at 10 03 41

## Test
I tried adding another test to `src/util/ajax.test.ts` but it is impossible to simulate a CORS with the `Nise` package's `FakeServer`, since a CORS error is not part of the server response but is due to the browser blocking it.

## Other info
- This makes no change to public API.
- this does not affect performance

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for taking the time to report this issue!
Can this be tested using integration tests in browser or unit tests?

File: src/util/ajax.ts [hunk 1]
```diff
@@ -295,3 +310,27 @@ export const getVideo = (urls: Array<string>): Promise<HTMLVideoElement> => {
...
+ * Check whether the provided error is due to CORS policy
+ */
+function isCorsError(error: unknown): boolean {
+    if (error instanceof TypeError) {
+        // Common CORS error messages can include:
+        // - "Failed to fetch"
+        // - "Network request failed"
+        // - "CORS error"
+        // - "Cross-Origin Request Blocked"
+        const corsErrorMessages = [
```
Comment (neodescis) [inline src/util/ajax.ts:324]: Two of these are pretty generic... are we certain they are limited to CORS?

Comment (jonathanlurie) [reply src/util/ajax.ts:324]: You're actually right, the message "Failed to fetch" can occur when there is a DNS issue and/or if the URL is malformed or invalid. I believe those case should also be dealt with the same way and still should throw an AJAXError, but I will adjust the messages and comments

Comment (jonathanlurie) [reply src/util/ajax.ts]: This is now updated.

**Comment by jonathanlurie** [comment]
> Thanks for taking the time to report this issue! Can this be tested using integration tests in browser or unit tests?

This is quite challenging to test since the fake server (`nise`) cannot simulate CORS issue or DNS resolution failures. I haven't find a way to test that yet.

File: src/util/ajax.ts [hunk 2]
```diff
@@ -155,7 +155,22 @@ async function makeFetchRequest(requestParameters: RequestParameters, abortContr
...
 
-    const response = await fetch(request);
+    // When the error is due to CORS policy, DNS issue or malformed URL, the fetch call does not resolve but throws a generic TypeError instead.
+    // It is preferable to throw an AJAXError so that the Map event "error" can catch it and still have
+    // access to the faulty url.
+    let response: Response;
+    try {
+        response = await fetch(request);
+    } catch (e) {
+        if (isNetworkError(e)) {
```
Comment (HarelM) [inline src/util/ajax.ts:165]: Why do we care if this is a network error then? If the missing information is that we want to get the url for the relevant fetch behavior, why not simply wrap any exception here with AJAXError?

**Comment by HarelM** [comment]
Yes, all the tests set `fetch to null`, so this can't be really tested.

File: src/util/ajax.ts (hunk 2 above)
Comment (jonathanlurie) [reply src/util/ajax.ts:165]: Yes, that's a good point, we could do that. I am not aware of all the possible scenarios in which `fetch` would throw, but if it does, that's fair to call it an AJAXError regardless of the reason and to make the faulty URL part of it.

**Comment by jonathanlurie** [comment]
For some weird reason, this PR is now coming with other things from the main branch. Closing it and reopening it clean.

**Comment by jonathanlurie** [comment]
Clean version here https://github.com/maplibre/maplibre-gl-js/pull/4931 , sorry for that!

**Pull request closed without merging**